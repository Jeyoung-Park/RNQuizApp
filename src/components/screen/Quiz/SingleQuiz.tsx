import * as React from 'react';
import { useEffect, useState } from 'react';
import {
	Text,
	View,
	StyleSheet,
	TouchableOpacity,
	SafeAreaView,
} from 'react-native';
import SimpleToast from 'react-native-simple-toast';
import { useSelector } from 'react-redux';
import { Quiz } from '../../../interface';
// import { Quiz } from '../../../interface';

interface SingleQuizProps {
	currentIndex: number;
	// setCurrentIndex: (param: number) => void;
}

const SingleQuiz = ({ currentIndex }: SingleQuizProps) => {
	console.log('currentIndex', currentIndex);

	const quizList = useSelector((state) => state.quiz.quizList.result?.results);

	const [currentQuiz, setCurrentQuiz] = useState<Quiz>();
	const [selections, setSelections] = useState<string[] | undefined>([]);
	const [choice, setChoice] = useState<string>();

	console.log('currentQuiz in SIngleQuiz, ', currentQuiz);

	useEffect(() => {
		setCurrentQuiz(quizList?.[currentIndex]);
	}, [currentIndex, quizList]);

	useEffect(() => {
		const randomNum = Math.floor(Math.random() * 4);
		console.log('randomNum, ', randomNum);
		console.log('incorrect answers before, ', currentQuiz?.incorrect_answers);
		const tempSelections = { ...currentQuiz }?.incorrect_answers;
		// tempSelections = [];
		tempSelections?.splice(randomNum, 0, currentQuiz?.correct_answer);
		console.log(
			'tempSelections, ',
			tempSelections,
			currentQuiz?.incorrect_answers,
			currentQuiz?.correct_answer,
		);
		setSelections(tempSelections);
	}, [currentQuiz]);

	useEffect(() => {
		if (choice === currentQuiz?.correct_answer) {
			SimpleToast.show('정답입니다.');
		} else {
			SimpleToast.show('오답입니다.');
		}
	}, [choice]);

	return (
		// eslint-disable-next-line @typescript-eslint/no-use-before-define
		<SafeAreaView style={styles.container}>
			<Text>
				Q{currentIndex + 1}. {currentQuiz?.category}
			</Text>
			<Text>{currentQuiz?.question}</Text>
			{selections?.map((item, index) => (
				<TouchableOpacity
					key={item + '' + index}
					onPress={() => {
						setChoice(item);
					}}
				>
					<Text>{item}</Text>
				</TouchableOpacity>
			))}
		</SafeAreaView>
	);
};

export default SingleQuiz;

const styles = StyleSheet.create({
	container: {
		flex: 1,

		// flexDirection: 'row',
		justifyContent: 'center',
		alignItems: 'center',
	},
});
