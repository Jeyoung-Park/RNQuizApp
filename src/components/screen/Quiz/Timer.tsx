/* eslint-disable @typescript-eslint/no-use-before-define */
import React, { useState, useEffect } from 'react';
import { Text, View, StyleSheet } from 'react-native';

interface TimerProps {
	isStart: boolean;
}

interface TimeProps {
	seconds: number;
	minutes: number;
}

const getTimeInFormat = (time: TimeProps) => {
	const { seconds, minutes } = time;
	const resultSeconds = seconds < 10 ? `0${seconds}` : seconds;
	const resultMinutes = minutes < 10 ? `0${minutes}` : minutes;

	return `${resultMinutes}:${resultSeconds}`;
};

const Timer = ({ isStart }: TimerProps) => {
	console.log('isStart in Timer, ', isStart);

	const [time, setTime] = useState<TimeProps>({
		seconds: 0,
		minutes: 0,
	});

	useEffect(() => {
		const advanceTime = () => {
			setTimeout(() => {
				let nSeconds = time.seconds;
				let nMinutes = time.minutes;

				nSeconds++;

				if (nSeconds > 59) {
					nMinutes++;
					nSeconds = 0;
				}

				if (isStart) {
					setTime({ seconds: nSeconds, minutes: nMinutes });
				}
			}, 1000);
		};
		advanceTime();
	}, [time]);

	return (
		<View style={styles.container}>
			<Text>{getTimeInFormat(time)}</Text>
		</View>
	);
};

export default Timer;

const styles = StyleSheet.create({
	container: {},
});
