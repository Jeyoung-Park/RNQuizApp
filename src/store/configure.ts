import { applyMiddleware, createStore } from 'redux';
import thunk from 'redux-thunk';
import reducer from '../models';

const configure = () => createStore(reducer, applyMiddleware(thunk));

export default configure;
