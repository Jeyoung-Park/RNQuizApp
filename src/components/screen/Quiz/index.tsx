import * as React from 'react';
import { useState } from 'react';
import { useEffect } from 'react';
import { Text, View, StyleSheet } from 'react-native';
import SimpleToast from 'react-native-simple-toast';
import { useDispatch, useSelector } from 'react-redux';
import { getQuizList_action } from '../../../models/quiz';
import SingleQuiz from './SingleQuiz';

interface QuizProps {
	// navigation?: any;
}

const Quiz = () => {
	const dispatch = useDispatch();

	const quizList = useSelector((state) => state.quiz.quizList.result?.results);

	const [currentIndex, setCurrentIndex] = useState(0);

	console.log('quizList in Quiz index, ', quizList);

	const getQuizList = async (numberOfQuiz: number = 10) => {
		try {
			const result = await dispatch(
				getQuizList_action({
					subPath: `?amount=${numberOfQuiz}&type=multiple`,
					params: null,
					data: null,
				}),
			);
			console.log('result in getQuizList, ', result);
		} catch (e) {
			console.warn('error in getQuizList, ', e);
			SimpleToast.show('퀴즈 목록을 불러오는 데 실패했습니다.');
		}
	};

	useEffect(() => {
		getQuizList(10);
		// eslint-disable-next-line react-hooks/exhaustive-deps
	}, []);

	return (
		// eslint-disable-next-line @typescript-eslint/no-use-before-define
		<View style={styles.container}>
			{/* <Text>Quiz</Text> */}
			<SingleQuiz
				currentIndex={currentIndex}
				setCurrentIndex={(param: number) => setCurrentIndex(param)}
			/>
		</View>
	);
};

export default Quiz;

const styles = StyleSheet.create({
	container: { flex: 1 },
});
