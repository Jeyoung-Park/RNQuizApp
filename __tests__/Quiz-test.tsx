import 'react-native';
import React from 'react';
import renderer from 'react-test-renderer';

import Quiz from '../src/components/screen/Quiz';
import { Provider } from 'react-redux';
import store from '../src/store';

jest.mock('react-native-simple-toast', () => ({
	SHORT: jest.fn(),
}));

it('snapshot test', () => {
	const tree = renderer
		.create(
			<Provider store={store}>
				<Quiz />
			</Provider>,
		)
		.toJSON();
	expect(tree).toMatchSnapshot();
});
