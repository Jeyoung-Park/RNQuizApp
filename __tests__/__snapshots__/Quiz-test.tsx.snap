// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`snapshot test 1`] = `
<RCTSafeAreaView
  emulateUnlessSupported={true}
  style={
    Object {
      "backgroundColor": "white",
      "flex": 1,
    }
  }
>
  <View
    style={
      Object {
        "alignItems": "center",
        "borderBottomColor": "gray",
        "borderBottomWidth": 1,
        "flexDirection": "row",
        "justifyContent": "space-between",
        "paddingHorizontal": 16,
        "paddingVertical": 20,
        "width": "100%",
      }
    }
  >
    <View
      accessible={true}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "alignItems": "center",
          "height": 24,
          "justifyContent": "center",
          "opacity": 1,
          "width": 24,
        }
      }
    >
      <Text
        allowFontScaling={false}
        style={
          Array [
            Object {
              "color": "black",
              "fontSize": 24,
            },
            undefined,
            Object {
              "fontFamily": "anticon",
              "fontStyle": "normal",
              "fontWeight": "normal",
            },
            Object {},
          ]
        }
      >
        î˜ 
      </Text>
    </View>
    <Text
      style={
        Object {
          "fontSize": 20,
        }
      }
    >
      Quiz
    </Text>
    <View
      style={
        Object {
          "alignItems": "center",
          "height": 24,
          "justifyContent": "center",
          "width": 24,
        }
      }
    />
  </View>
  <RCTSafeAreaView
    emulateUnlessSupported={true}
    style={
      Object {
        "alignItems": "center",
        "flex": 1,
        "paddingHorizontal": 16,
      }
    }
  >
    <View
      style={
        Object {
          "alignItems": "center",
          "flexDirection": "row",
          "justifyContent": "space-between",
          "marginTop": 20,
          "width": "100%",
        }
      }
    >
      <Text>
        Q
        1
      </Text>
      <View
        style={Object {}}
      >
        <Text>
          00:00
        </Text>
      </View>
      <Text>
        0
        /
      </Text>
    </View>
    <View
      style={
        Object {
          "flex": 1,
          "justifyContent": "center",
        }
      }
    >
      <Text>
        
      </Text>
    </View>
    <View
      style={
        Object {
          "flex": 2,
          "justifyContent": "center",
        }
      }
    >
      <Text>
        
      </Text>
    </View>
    <View
      style={
        Object {
          "flex": 3,
          "justifyContent": "center",
        }
      }
    />
  </RCTSafeAreaView>
  <View
    style={
      Object {
        "alignItems": "center",
        "flexDirection": "row",
        "height": 50,
        "justifyContent": "flex-end",
        "paddingRight": 20,
      }
    }
  />
</RCTSafeAreaView>
`;
